chain-tx-enclave-next/tdbe/app-runner/src/main.rs

```diff
@@ -0,0 +1,180 @@
+    /// Transaction IDs for testing
+    #[structopt(long = "txids")]
+    pub txids: Vec<String>,
+    /// Use dummy signature in testing
+    #[structopt(short = "t", long = "test")]
+    pub test: bool,
+}
+
+impl Options {
+    /// Validates provided command line options
+    fn validate(&self) {
+        match (self.tdbe_address.as_ref(), self.tdbe_dns_name.as_ref()) {
+            (Some(_), None) | (None, Some(_)) => {
+                panic!("Either both, `tdbe-address` and `tdbe-dns-name` should be provided or none of them should be provided");
+            }
+            _ => {
+                // Options are valid
+            }
+        }
+    }
+}
+
+#[allow(clippy::type_complexity)]
+impl UsercallExtension for Options {
+    fn connect_stream<'future>(
+        &'future self,
+        addr: &'future str,
+        _local_addr: Option<&'future mut String>,
+        _peer_addr: Option<&'future mut String>,
+    ) -> Pin<Box<dyn Future<Output = io::Result<Option<Box<dyn AsyncStream>>>> + 'future>> {
+        async fn connect_stream_inner(
+            this: &Options,
+            addr: &str,
+        ) -> io::Result<Option<Box<dyn AsyncStream>>> {
+            match addr {
+                "init" => {
+                    let transaction_ids = parse_transaction_ids(&this.txids);
+
+                    let tdbe_config = TdbeConfig {
+                        transaction_ids,
+                        tdbe_dns_name: this.tdbe_dns_name.clone(),
+                    };
+
+                    let mut stream = Cursor::new(Vec::new());
+                    tdbe_config
+                        .write_to(&mut stream)
+                        .expect("Unable to write initial configuration to `Cursor`");
+
+                    Ok(Some(Box::new(stream)))
+                }
+                "zmq" => {
+                    let zmq_helper = ZmqHelper::new(&this.zmq_conn_str)?;
+                    Ok(Some(Box::new(zmq_helper)))
+                }
+                "ra-sp-server" => {
+                    let stream = TcpStream::connect(&this.sp_address).await?;
+                    Ok(Some(Box::new(stream)))
+                }
+                "tdbe" => match this.tdbe_address {
+                    Some(ref tdbe_address) => {
+                        let stream = TcpStream::connect(tdbe_address).await?;
+                        Ok(Some(Box::new(stream)))
+                    }
+                    None => {
+                        log::error!("Attempting to connect to another TDBE server but no address provided at startup.");
+                        Ok(None)
+                    }
+                },
+                _ => Ok(None),
+            }
+        }
+
+        Box::pin(connect_stream_inner(self, addr))
+    }
+
+    fn bind_stream<'future>(
+        &'future self,
+        addr: &'future str,
+        _local_addr: Option<&'future mut String>,
+    ) -> Pin<Box<dyn Future<Output = io::Result<Option<Box<dyn AsyncListener>>>> + 'future>> {
```

Comment:
how about using `type` to simplify this line?

Reply:
We don't use this type anywhere else. Its just the type returned by this trait: https://docs.rs/enclave-runner/0.4.0/enclave_runner/usercalls/trait.UsercallExtension.html